Cython/Compiler/ExprNodes.py

```diff
@@ -1279,6 +1279,11 @@ def coerce_to(self, dst_type, env):
             error(self.pos, "Cannot assign None to %s" % dst_type)
         return super().coerce_to(dst_type, env)
 
+    def analyse_as_type(self, env):
+        if not env.in_c_type_context:
+            return py_none_type
+        return None
```

Comment:
Let's avoid the negation here.
```suggestion
    def analyse_as_type(self, env):
        if env.in_c_type_context:
            return None
        return py_none_type
```